### PR TITLE
Entity signing

### PR DESCRIPTION
**Pull request by rllola**
Closes https://github.com/ZondaX/ledger-oasis/issues/12

### CODE REVIEW

File: src/lib/parser_txdef.h [hunk 1]
```diff
@@ -99,7 +101,21 @@ typedef struct {
...
+} oasis_entity_t;
+
+typedef struct {
+
+    union {
+        oasis_tx_t oasis_tx;
+        oasis_entity_t oasis_entity;
+    };
+
+    bool is_tx;
```
Comment (jleni) [inline src/lib/parser_txdef.h]: an `enum` would be better in case we need to extend this later on

Comment (jleni) [reply src/lib/parser_txdef.h]: also it is a good idea that the first element ( 0 ) in the enum is `TYPE_UNKNOWN` and we later reject that during validation. When the parser is initialized, we set the `type` to `TYPE_UNKNOWN`

File: src/lib/parser_txdef.h [hunk 2]
```diff
@@ -99,7 +101,21 @@ typedef struct {
...
+    size_t nodes_length;
+    bool allow_entity_signed_nodes;
+} oasis_entity_t;
+
+typedef struct {
+
+    union {
+        oasis_tx_t oasis_tx;
+        oasis_entity_t oasis_entity;
+    };
```
Comment (jleni) [inline src/lib/parser_txdef.h]: we may need to name the union.. maybe something like this?
```
    union {
        oasis_tx_t tx;
        oasis_entity_t entity;
    } oasis;
```
that way later it is possible to MEMZERO the complete union which is safer

File: src/lib/parser_impl.c [hunk 3]
```diff
@@ -394,46 +392,104 @@ __Z_INLINE parser_error_t _readMethod(parser_tx_t *v, CborValue *value) {
...
+    if (cbor_value_at_end(&it)) {
+        return parser_unexpected_buffer_end;
+    }
+
+    if (!cbor_value_is_map(&it)) {
+        return parser_unexpected_type;
+    }
+
+    // ENTITY OR TX ?
+    // REVIEW: Might be a better way to do it ?
```
Comment (jleni) [inline src/lib/parser_impl.c]: Not really apart from matching fields that we know should be there.
Later when parsing each specific case, we need to be sure there are no extra fields.

File: src/lib/parser.c [hunk 4]
```diff
@@ -372,3 +363,62 @@ parser_error_t parser_getItem(parser_context_t *ctx,
...
+    MEMZERO(outKey, outKeyLen);
+    MEMZERO(outVal, outValLen);
+    snprintf(outKey, outKeyLen, "?");
+    snprintf(outVal, outValLen, " ");
+
+    if (displayIdx < 0 || displayIdx >= parser_getNumItems(ctx)) {
+        return parser_no_data;
+    }
+
+    // REVIEW: can't verify type in C
```
Comment (jleni) [inline src/lib/parser.c]: not a reliable one. I think the suggestion of using an enum to specify the struct embeded in the union would be an acceptable solution

**Review by jleni — CHANGES_REQUESTED** [review]

File: src/lib/parser_impl.c [hunk 5]
```diff
@@ -589,8 +657,32 @@ parser_error_t _getCommissionBoundStepAtIndex(parser_context_t *c, parser_tx_t *
...
+                                    c->offset,
+                                    &v->parser,
+                                    &it));
+
+    if (cbor_value_at_end(&it)) {
+        return parser_unexpected_buffer_end;
+    }
+
+    CborValue nodesContainer;
+    CHECK_CBOR_ERR(_getBoundsContainer(&it, &nodesContainer));
```
Comment (jleni) [inline src/lib/parser_impl.c]: Is this correct? It seems to be the wrong function to call here.

**Review by jleni — CHANGES_REQUESTED** [review]

**Comment by jleni** [comment]
This PR seems to be addressing https://github.com/ZondaX/ledger-oasis/issues/20 too, right?

File: src/lib/parser_impl.c (hunk 5 above)
Comment (rllola) [reply src/lib/parser_impl.c]: yeah that was a mistake. It should be resolved now.

**Comment by rllola** [comment]
> This PR seems to be addressing [ZondaX/ledger-oasis#20](https://github.com/ZondaX/ledger-oasis/issues/20) too, right?

Yes it will fix ZondaX/ledger-oasis#20

**Pull request merged**